### PR TITLE
[xpu][feature]Enable SYCL Native Fast Math Support in `NumericUtils.h`

### PR DESCRIPTION
**Pull request by chunhuanMeng**
Updates several mathematical utility functions in `aten/src/ATen/NumericUtils.h` to add support for SYCL device compilation. Specifically, it introduces fast approximations for `exp`, `log`, `log1p`, and `tan` using SYCL's native math functions when compiling for SYCL devices. This ensures optimal performance and compatibility across CUDA, HIP, and SYCL backends.

Device-specific math optimizations:

* Updated the `exp`, `log`, `log1p`, and `tan` functions to use `sycl::native::exp`, `sycl::native::log`, and `sycl::native::tan` for fast approximations when compiling for SYCL devices. This matches the approach already used for CUDA and HIP devices, improving performance and consistency across different hardware backends.

cc @gujinghui @EikanWang @fengyuan14 @guangyey

### CODE REVIEW

**Review by guangyey — APPROVED** [review]

**Review by EikanWang — CHANGES_REQUESTED** [review]
Test cases are required for different data types.

File: aten/src/ATen/NumericUtils.h [hunk 1]
```diff
@@ -135,6 +135,9 @@ C10_HOST_DEVICE inline T exp(T x) {
 #if defined(__CUDA_ARCH__) || defined(__HIP_ARCH__)
   // use __expf fast approximation for peak bandwidth
   return __expf(x);
+#elif defined(__SYCL_DEVICE_ONLY__)
+  // use native::exp fast approximation for peak bandwidth
```
Comment (gujinghui) [inline aten/src/ATen/NumericUtils.h:139]: The comments confused me. Our motivation to replace std with sycl::native is to achieve better performance or high precision?

Comment (chunhuanMeng) [reply aten/src/ATen/NumericUtils.h:139]: The motivation is to achieve better performance, similar to CUDA's approach.

Comment (gujinghui) [reply aten/src/ATen/NumericUtils.h:139]: Looks like CUDA has same comments. Here peak bandwidth means better performance, instead high precision?

Comment (chunhuanMeng) [reply aten/src/ATen/NumericUtils.h:139]: You're right. In this context, “peak bandwidth” means better performance using fast approximations rather than higher precision.

File: aten/src/ATen/NumericUtils.h [hunk 2]
```diff
@@ -135,6 +135,9 @@ C10_HOST_DEVICE inline T exp(T x) {
...
+  return sycl::native::exp(x);
```
Comment (EikanWang) [inline aten/src/ATen/NumericUtils.h:140]: @jianyizh , have we checked with the sycl community that `sycl::native::{math_func}` is the fast math version?

Comment (chunhuanMeng) [reply aten/src/ATen/NumericUtils.h:140]: @EikanWang https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_native_precision_math_functions

**Comment by EikanWang** [comment]
@chunhuanMeng @guangyey , could you help check the two failures?

**Comment by chunhuanMeng** [comment]
> @chunhuanMeng @guangyey , could you help check the two failures?

@EikanWang These two failures are not related to this patch:.

**Comment by guangyey** [comment]
test/test_foreach.py::TestForeachXPU::test_foreach_max_empty_tensor_error__foreach_max_xpu_bfloat16 is fixed in https://github.com/pytorch/pytorch/pull/183722

**Comment by guangyey** [comment]
Yes, these failure are all unrelated.

**Comment by EikanWang** [comment]
@claude is the failure related?

**Review by EikanWang — APPROVED** [review]

**Comment by EikanWang** [comment]
@chunhuanMeng , although the changes have been covered by other unit test cases, some dedicated unit test cases in test_xpu.py are preferred.

**Comment by guangyey** [comment]
@janeyx99 does this code change make sense to you? I'd appreciate your thoughts when you have a chance.

**Comment by janeyx99** [comment]
@guangyey Is there anything in particular you'd like my opinion on? I took a look at the change and it looks reasonable. If you're asking whether it's okay to start adding `#ifdef SYCL_...` branching everywhere in aten, then that is a larger question for @albanD.

I did not look look into the test case, and I don't see benchmarks showing the speedups vs the default, but I'd defer to XPU experts on those parts.

**Comment by guangyey** [comment]
Hi @janeyx99, Thanks for taking a look and for the feedback!
The main thing I was looking for was whether the overall approach and code change seemed reasonable, so it's good to hear that it does.
Your feedback is very helpful. We'll add benchmark results comparing against the default path so the performance impact is clearer.
Regarding the `#ifdef __SYCL...` usage, we are conservative about introducing `#ifdef __SYCL...` branches into ATen, Our principle is to keep them aligned with the existing CUDA implementation, and only introduce them when there is a clear, measurable performance benefit. I'll ask @albanD  for his thoughts on whether it makes sense from the overall design perspective.

**Comment by guangyey** [comment]
@chunhuanMeng, could you please update the performance comparison in the PR description.

**Comment by chunhuanMeng** [comment]
| Op | dtype | size | original(us) | now(us) | speedup ratio |
|---|---|---|---|---|---|
| **cauchy_** | float32 | 10M | 91.4 | 54.4 | **1.68x** |
| **cauchy_** | float16 | 10M | 94.4 | 55.8 | **1.69x** |
| **cauchy_** | bfloat16 | 10M | 108.3 | 55.0 | **1.97x** |
| **log_normal_** | float32 | 10M | 93.5 | 69.8 | **1.34x** |
| **log_normal_** | float16 | 10M | 110.3 | 74.6 | **1.48x** |
| **log_normal_** | bfloat16 | 10M | 107.8 | 78.3 | **1.38x** |

File: test/test_xpu.py [hunk 3]
```diff
@@ -3042,6 +3021,60 @@ def test_two_graphs(
                 self.assertNotEqual(new_data_ptr, old_data_ptr)
 
 
+@unittest.skipIf(not TEST_XPU, "XPU not available, skipping tests")
```
Comment (albanD) [inline test/test_xpu.py:3024]: Two questions here?
1. We only test distributions because the base ops that we changed are already covered by OpInfo?
2. Why are the distributions tests that are device generic are not enough?

**Review by albanD — APPROVED** [review]
Questions to improve testing mostly.
If you really think this is enough, go for it.

File: test/test_xpu.py (hunk 3 above)
Comment (chunhuanMeng) [reply test/test_xpu.py:3024]: Hi @albanD, thanks for the approval and the questions!

**Q1: We only test distributions because the base ops that we changed are already covered by OpInfo?**

Not exactly. The functions we changed (`at::exp`, `at::log`, `at::log1p`, `at::tan` in NumericUtils.h) are device-inline helper functions that are **only** called from `TransformationHelper.h`, which is used exclusively by distribution kernels (`cauchy_`, `log_normal_`, `exponential_`, `geometric_`).

The PyTorch ops `torch.exp`, `torch.log`, `torch.tan` (which are covered by OpInfo) go through a completely different code path (UnaryOps kernels) and do **not** call NumericUtils.h functions. So OpInfo tests don't exercise our change at all — only distribution ops do.

**Q2: Why are the device-generic distribution tests not enough?**

The existing device-generic distribution tests do cover basic correctness on XPU. The dedicated tests added here provide additional validation specific to the fast-math approximation:
- They verify precision bounds (e.g., median accuracy within tolerance) to catch potential precision regressions from `sycl::native::*` approximations
- They test across multiple dtypes (float32/float16/bfloat16) since native fast math precision characteristics can vary by dtype
- They serve as a regression guard: if a future compiler change degrades `sycl::native::tan/exp/log` precision, these tests would catch it
